src/execution/flattenAsyncIterable.ts

```diff
@@ -0,0 +1,50 @@
+import { isAsyncIterable } from '../jsutils/isAsyncIterable';
+
+type AsyncIterableOrGenerator<T> =
+  | AsyncGenerator<T, void, void>
+  | AsyncIterable<T>;
+
+/**
+ * Given an AsyncIterable that could potentially yield other async iterators,
+ * flatten all yielded results into a single AsyncIterable
+ */
+export function flattenAsyncIterable<T, AT>(
+  iterable: AsyncIterableOrGenerator<T | AsyncIterableOrGenerator<AT>>,
+): AsyncGenerator<T | AT, void, void> {
+  const iteratorMethod = iterable[Symbol.asyncIterator];
+  const iterator: any = iteratorMethod.call(iterable);
+  let iteratorStack: Array<AsyncIterator<T>> = [iterator];
+
+  async function next(): Promise<IteratorResult<T | AT, void>> {
+    const currentIterator = iteratorStack[0];
+    if (!currentIterator) {
+      return { value: undefined, done: true };
+    }
+    const result = await currentIterator.next();
```

Comment:
flattenAsyncIterable is buggy if you call `next` multiple times in parallel.
https://github.com/graphql/graphql-js/pull/3709

Reply:
@robrichard It looks like you rewrote flattenAsyncIterable as a simple async generator. Note that that has the problem that you can't stop (`return`) the generator while it's blocked on reading (ie anywhere other than at the `yield`. That was causing issues in one of the subscribe tests for me when I tried it that way, and in fact I see that it's failing for you too (ie CI is failing). I think this is a fundamental issue with using async generator functions that wrap other async iterators (see eg https://github.com/tc39/proposal-async-iteration/issues/126)

In https://github.com/graphql/graphql-js/pull/3703 I do a few things — main motivation is simplifying/explicit-ifying the types used, but I also think I fix flattenAsyncIterable without introducing the `return` problem. Though I'll be honest, all that stuff around stopping generators is weird and I can believe I haven't implemented it quite correctly either.

Reply:
this was fixed in https://github.com/graphql/graphql-js/pull/3710